zip-0317.rst

```diff
@@ -8,9 +8,308 @@
+
+* :math:`marginal\_fee = 5000`;
+* :math:`grace\_actions = 2`.
+
+It is not a consensus requirement that fees follow this formula; however,
+wallets SHOULD create transactions that pay this fee, in order to reduce
+information leakage, unless overridden by the user.
+
+Rationale for logical actions
+'''''''''''''''''''''''''''''
+
+The intention is to make the fee paid for a transaction depend on its
+impact on the network, without discriminating between different protocols
+(Orchard, Sapling, or transparent). The impact on the network depends on
+the numbers of inputs and outputs.
+
+A previous proposal used :math:`inputs + outputs` instead of logical actions.
+This would have disadvantages Orchard transactions, as a result of an
+Orchard Action combining an input and an output. The effect of this
+combining is that Orchard requires padding of either inputs or outputs
+to ensure that the number of inputs and outputs are the same. Usage of
+Sapling and transparent protocols does not require this padding, and
+so this could have effectively discriminated against Orchard.
+
+Rationale for the chosen parameters
+'''''''''''''''''''''''''''''''''''
+
+Grace Actions
+~~~~~~~~~~~~~
+
+**Why not just charge per-action, without a grace window?**
+
+* This ensures that there is no penalty to padding a 1-action
+  transaction to a 2-action transaction. Such padding is desirable
+  to reduce information leakage from input and output arity, and
+  is the standard approach used by `zcashd` and the mobile SDK
+  transaction builder.
+* Without a grace window, an input with value below the marginal
+  fee would never be worth including in the resulting transaction.
+  With a grace window, an input with value below :math:`marginal\_fee`
+  *is* worth including, if a second input is available that covers
+  both the primary output amount and the conventional transaction
+  fee.
+
+**Why a grace window of 2?**
+
+A 1-in, 2-out (or 2-action) transaction is the smallest possible
+transaction that permits both an output to a recipient, and a
+change output. However, as stated above, `zcashd` and the mobile
+SDK transaction builder will pad the number of inputs to at least 2.
+
+Let :math:`min\_actions` be the minimum number of logical actions
+that can be used to execute economically relevant transactions that
+produce change. Due to the aforementioned padding, :math:`min\_actions = 2`.
+
+Having a grace window size greater than :math:`min\_actions` would
+increase the cost to create such a minimal transaction. If the
+cost for a minimal transaction is bounded above by :math:`B`, then
+possible choices of :math:`marginal\_fee` are bounded above by
+:math:`B / max(min\_actions, grace\_actions)`. Therefore, the
+optimal choice of :math:`grace\_actions` to maximize the cost of
+denial-of-service attacks that use many logical actions, without
+imposing an undue penalty on minimal transactions, is
+:math:`grace\_actions = min\_actions = 2`.
```

Comment:
```suggestion
```suggestion
cost we believe that users will tolerate for a minimal transaction
is :math:`B`, then possible choices of :math:`marginal\_fee` are
bounded above by :math:`B / \max(min\_actions, grace\_actions)`.
Therefore, the optimal choice of :math:`grace\_actions` to maximize
the cost of denial-of-service attacks that use many logical actions,
without imposing an undue penalty on minimal transactions, is
:math:`grace\_actions = min\_actions = 2`.
```